### PR TITLE
Implement New Model

### PR DESCRIPTION
**Pull request by case141**
Added new classes to source package created in the current model:
- Title
- Type
- Detail
- Source

### CODE REVIEW

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -236,7 +236,7 @@
     -->
 
     <!-- Checks that the order of at-clauses follows the tagOrder default property value order.
-         @author, @version, @param, @return, @throws, @exception, @see, @since, @serial, @serialField, @serialData, @deprecated
+         @author, @version, @param, @return, @throws, @exceptions, @see, @since, @serial, @serialField, @serialData, @deprecated
```
Comment (fterh) [inline config/checkstyle/checkstyle.xml]: Why is this changed?

File: database.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
 Algorithms Wikipedia y/Website d/Algorithm basics from wikipedia t/Algorithms t/Wikipedia
 Telecomms: A New Insight y/Journal d/Better way to manage telecommunications t/Telecommunications
 some title y/some type d/details of thing t/some tag
+ALGORITHMS ARTICLE y/website d/something  t/
```
Comment (fterh) [inline database.txt]: Is this intentional?

File: src/main/java/seedu/address/commons/exceptions/IllegalValueException.java [hunk 3]
```diff
@@ -13,7 +13,7 @@ public IllegalValueException(String message) {
 
     /**
      * @param message should contain relevant information on the failed constraint(s)
-     * @param cause of the main exception
+     * @param cause of the main exceptions
```
Comment (fterh) [inline src/main/java/seedu/address/commons/exceptions/IllegalValueException.java]: Why are all the singular exception changed to plural?

File: src/main/java/seedu/address/model/source/Detail.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
+package seedu.address.model.source;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+public class Detail {
+
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java:11]: Does the styleguide say anything about 1 or 2 line breaks here?

File: src/main/java/seedu/address/model/source/Detail.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    public static final String MESSAGE_CONSTRAINTS = "Details should be alphanumeric";
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java:12]: Maybe rename MESSAGE_CONSTRAINTS to DETAIL_CONSTRAINT_MESSAGE to make it clearer

File: src/main/java/seedu/address/model/source/Detail.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    public static final String VALIDATION_REGEX = "\\p{Alnum}+";
+
+    public final String detail;
+
+    /**
+     * Constructs a {@code Tag}.
+     *
+     * @param detail A valid tag name.
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java]: Why not rename the parameter from `detail` to `name` then?

File: src/main/java/seedu/address/model/source/Detail.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public Detail(String detail) {
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java:22]: I'd suggest renaming `String detail` to `String name`

File: src/main/java/seedu/address/model/source/Detail.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+public class Detail {
+
+
+    public static final String MESSAGE_CONSTRAINTS = "Details should be alphanumeric";
+    public static final String VALIDATION_REGEX = "\\p{Alnum}+";
+
+    public final String detail;
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java:15]: Maybe rename this to `name`

File: src/main/java/seedu/address/model/source/Detail.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+     * @param detail A valid tag name.
+     */
+    public Detail(String detail) {
+        requireNonNull(detail);
+        checkArgument(isValidDetail(detail), MESSAGE_CONSTRAINTS);
+        this.detail = detail;
+    }
+
+    /**
+     * Returns true if a given string is a valid tag name.
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java]: You can use the `@return` clause.

See: https://www.oracle.com/technetwork/articles/java/index-137868.html

File: src/main/java/seedu/address/model/source/Detail.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public static boolean isValidDetail(String test) {
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java:31]: 1. Maybe rename `test` to `testString` or `testSubject` to make it more descriptive?
2. Should this method be public? If it's an internal utility method maybe make it private.
3. Do you want to be more defensive and ensure `test != null` to prevent null pointer exception?

File: src/main/java/seedu/address/model/source/Detail.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        return test.matches(VALIDATION_REGEX);
+    }
+
+    @Override
+    public boolean equals(Object other) {
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java:36]: Food for thought: if I subclass `Detail`, should the `equals` method still return true? Currently, it will.

File: src/main/java/seedu/address/model/source/Detail.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+                && detail.equals(((Detail) other).detail)); // state check
+    }
+
+    @Override
+    public int hashCode() {
+        return detail.hashCode();
+    }
+
+    /**
+     * Format state as text for viewing.
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java:48]: "Format" should be "Formats"

File: src/main/java/seedu/address/model/source/Detail.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public String toString() {
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Detail.java:51]: Should there be @Override annotation for this?

File: src/main/java/seedu/address/model/source/Source.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+public class Source {
+
+    private final Title title;
+    private final Type type;
+    private final Detail detail;
+
+    private final Set<Tag> tags = new HashSet<>();
+
+    public Source(Title title, Type type, Detail detail){
+
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Source.java]: Unnecessary line break

File: src/main/java/seedu/address/model/source/Source.java [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        this.title = title;
+        this.type = type;
+        this.detail = detail;
+
+    }
+
+    public Title getTitle() { return title; }
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Source.java]: Don't put same line, code style violation

File: src/main/java/seedu/address/model/source/Source.java [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+    public Type getType() { return type; }
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Source.java]: Same

File: src/main/java/seedu/address/model/source/Source.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+    public Detail getDetail() { return detail; }
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Source.java]: Same

File: src/main/java/seedu/address/model/source/Source.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Returns an immutable tag set, which throws {@code UnsupportedOperationException}
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Source.java:45]: Use `@returns` clause

File: src/main/java/seedu/address/model/source/Source.java [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+     * if modification is attempted.
+     */
+    public Set<Tag> getTags() {
+        return Collections.unmodifiableSet(tags);
+    }
+
+    /**
+     * Returns true if both sources of the same title have at least one other identity field that is the same.
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Source.java:53]: Use `@returns` clause

File: src/main/java/seedu/address/model/source/Source.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+     * This defines a weaker notion of equality between two sources.
+     */
+    public boolean isSameSource(Source otherSource) {
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Source.java:56]: Why do we need this method?

File: src/main/java/seedu/address/model/source/Source.java [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+            return true;
+        }
+
+        return otherSource != null
+                && otherSource.getTitle().equals(getTitle())
+                && (otherSource.getType().equals(getType()) || otherSource.getDetail().equals(getDetail()));
+    }
+
+    /**
+     * Returns true if both sources have the same identity and data fields.
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Source.java:67]: Use `@returns` clause

File: src/main/java/seedu/address/model/source/Title.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+    public Title(String title) {
+        requireNonNull(title);
+        checkArgument(isValidTitle(title), MESSAGE_CONSTRAINTS);
+        this.title = title;
+    }
+
+    /**
+     * Returns true if a given string is a valid name.
+     */
+    public static boolean isValidTitle(String test) {
```
Comment (fterh) [inline src/main/java/seedu/address/model/source/Title.java:36]: This method looks similar to the one in Tag.

Maybe create a parent class to refactor this code to apply DRY?

**Review by fterh — COMMENTED** [review]
Why are all the "exception" renamed to "exceptions"? 😕
